### ISSUE TITLE
Internal compiler error on nightly: 'index out of bounds'

### ISSUE THREAD

**Issue by avl** [rust-lang/rust#48923]
I just got an internal compiler error. This happened after introducing a custom derive macro and using it on several structs. I'll try to narrow it down more, and will comment on this report if I make any progress.

## Meta

rustc 1.26.0-nightly (2789b067d 2018-03-06)
binary: rustc
commit-hash: 2789b067da2ac921b86199bde21dd231ace1da39
commit-date: 2018-03-06
host: x86_64-unknown-linux-gnu
release: 1.26.0-nightly
LLVM version: 6.0

Backtrace:

thread 'rustc' panicked at 'index out of bounds: the len is 241450 but the index is 241450', /checkout/src/libcore/slice/mod.rs:871:14
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at libstd/sys_common/backtrace.rs:59
             at libstd/panicking.rs:207
   3: std::panicking::default_hook
             at libstd/panicking.rs:223
   4: core::ops::function::Fn::call
   5: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:403
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:349
   7: rust_begin_unwind
             at libstd/panicking.rs:325
   8: core::panicking::panic_fmt
             at libcore/panicking.rs:72
   9: core::panicking::panic_bounds_check
             at libcore/panicking.rs:58
  10: rustc::dep_graph::graph::DepGraph::with_task_impl
  11: rustc::ty::maps::<impl rustc::ty::maps::queries::crate_disambiguator<'tcx>>::force
  12: rustc::ty::maps::<impl rustc::ty::maps::queries::crate_disambiguator<'tcx>>::try_get
  13: rustc::ty::maps::TyCtxtAt::crate_disambiguator
  14: rustc::ty::maps::on_disk_cache::OnDiskCache::compute_cnum_map::{{closure}}::{{closure}}
  15: <std::collections::hash::map::HashMap<K, V, S> as core::iter::traits::Extend<(K, V)>>::extend
  16: rustc::dep_graph::graph::DepGraph::with_ignore
  17: rustc::ty::maps::on_disk_cache::OnDiskCache::load_diagnostics
  18: rustc::dep_graph::graph::DepGraph::try_mark_green
  19: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::try_mark_green_and_read
  20: rustc::ty::maps::<impl rustc::ty::maps::queries::target_features_enabled<'tcx>>::try_get
  21: rustc::ty::maps::TyCtxtAt::target_features_enabled
  22: <rustc::hir::check_attr::CheckAttrVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_item
  23: rustc::hir::check_attr::check_crate
  24: rustc::ty::context::TyCtxt::create_and_enter
  25: rustc_driver::driver::compile_input
  26: rustc_driver::run_compiler

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.26.0-nightly (2789b067d 2018-03-06) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental -C target-cpu=native --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `fractal`.

To learn more, run the command again with --verbose.

**Comment by avl** [comment]
Using:

CARGO_INCREMENTAL=0 cargo build

instead of just cargo build, seems to resolve the problem.

**Comment by avl** [comment]
My program isn't correct, so I guess the problem is with some kind of diagnostics.

**Comment by avl** [comment]
Ok, now that I've fixed the compilation errors, the program compiles ok with CARGO_INCREMENTAL=0 . It even appears to run correctly, but it still cannot be recompiled with CARGO_INCREMENTAL=1 . 

It can be compiled with CARGO_INCREMENTAL=0
It can be compiled once (after "cargo clean"), with CARGO_INCREMENTAL=1
Subsequent  compiles with CARGO_INCREMENTAL=1 crash, even after a whitespace-only change.

**Comment by avl** [comment]
Ok, another lead: I was using my 'savefile'  crate (which is on crates.io) with its custom-derive macro to derive an implementation of a trait called Serialize .I was also, in the crate which would not compile, using serde to derive its trait called Serialize. These are different traits. Perhaps the incremental compilation could not keep these two apart somehow? My immediate problem is solved, since I was trying out both serde and savefile, and never intended to actually use both simultaneously.

**Comment by Ameobea** [comment]
I've got the same problem in a crate I'm developing.  I'm also using `serde` and `serde-derive`.  The error only seems to occur after a compilation fails due to type error or similar.  Deleting the `target/debug/incremental` directory allow the crate to compile successfully.

**Comment by ExpHP** [comment]
This is tagged `A-type-system`, but #49065 shows another instance of this bug that occurred during trans.  The common factor between the two is an index error in `DepGraph`.

**Comment by michaelwoerister** [comment]
This looks indeed like a problem with incremental compilation.

**Comment by ExpHP** [comment]
@jjl @avl @Ameobea since the crash seems to at least be deterministic, can anyone share the code that reproduces the crash? (minimizing it would be a huge help as well, but simply having anything would be great at this point)

Barring that, you can also help by testing other nightlies to pinpoint when the problem first started occurring. (you can do `rustup override set nightly-YYYY-MM-DD` to use a nightly from a specific day)

**Comment by nikomatsakis** [comment]
triage: P-high

**Comment by jjl** [comment]
My code is not open source. I'm willing to share the code privately, but not publically on github, if we can come up with a way to do that.

**Comment by avl** [comment]
I'm trying to minimize the program which causes the crash for me. I have come to where it seems to work if I remove a dependency I have upon another crate in the same workspace. Does the incremental compilation work "per crate", or is the cached intermediaries shared between crates? Is it reasonable that the bug could be triggered by code patterns in the combination of two crates where one depends on the other?

**Comment by avl** [comment]
Hmm,  minimizing is a bit tricky. After the error has been triggered once, almost any input will yield an ICE. For instance, completely clearing lib.rs in the offending crate would still trigger ICE. However, removing dependencies from Cargo.toml would make it work again. Maybe the intermediary state/cache is thrown away when changes are detected in Cargo.toml?

**Comment by avl** [comment]
Ok, I'm making a lot of headway now :-) . Will have a minimal example here on github shortly.

**Comment by avl** [comment]
Okay, actually, this program triggers the error:

#[macro_use]
extern crate serde_derive;

#[derive(Serialize)]
pub struct X {
    pub x: f32,
}

#[derive(Serialize)]
pub struct Y {
    pub y: f32,
}

If I use the following Cargo.toml:

[package]
name = "fractal"
version = "0.1.0"
authors = ["Anders Musikka <you@example.com>"]

[dependencies]
serde="1.0.28"
serde_derive="1.0.28"
savefile-derive="0.2.5"

**Comment by avl** [comment]
Minimal example, including a convenient(?) crash.sh script:
https://github.com/avl/rust_ice_48923

**Comment by ExpHP** [comment]
How fascinating that even `extern crate savefile_derive` is not necessary!  Indeed, I can confirm that it consistently crashes on recompile when `savefile-derive` is listed under `[dependencies]`, and that it can be repeatedly compiled without issue if the dependency is removed.

It sounds to me like you are on to something about the `derive_Serialize`.  I wonder if the other crashing projects might have multiple versions of `serde_derive` linked in?

**Comment by avl** [comment]
Yeah, I thought that too! I (as a layperson) was also surprised that mentioning a crate in Cargo.toml could have an effect on compilation. Having multiple traits called Serialize is a bit suspicious, and having multiple serde versions seems like something that could the same situation to appear.

Hmm, but thinking about this a bit more: The savefile_derive crate is just a custom-derive macro crate. It doesn't actually define any traits, and it doesn't depend on the savefile crate which does define Serialize. So maybe it's actually proc_macro for a trait name that is the culprit, not the actual trait name itself (although these two are of course very related).

**Comment by avl** [comment]
Ok, so I proceeded a bit further. If I reduce the savefile-derive crate 'lib.rs' into:

#![feature(proc_macro)]
extern crate proc_macro;

#[proc_macro_derive(Savefile, attributes(versions, versions_as, default_val, default_fn))]
pub fn savefile(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    panic!("Not implemented")
}

The problem still appears. So we can eliminate the theory of multiple Serialize-traits being the culprit. Instead it seems like simply linking to a custom-derive crate at all can trigger the problem.

**Comment by ExpHP** [comment]
Can you do that on the `rust_issues_48923` repo?  (you can put a minimized `savefile-derive` in that source tree and depend on it with `savefile-derive = { path = "./savefile-derive" }`)

I tried repeating your experiment and am not able to reproduce it.  Minimizing `savefile_derive` down to just that (with the `Cargo.toml.orig` from crates.io) and depending on it in `rust_issues_48923` did not cause any crashes for me.  (however, it also continued to build fine even if I changed the `proc_macro_derive` to `Serialize`, so your conclusion may still be correct)

**Comment by RandomInsano** [comment]
I've hit a very similar problem on nightly while building the `cc` crate as a sub-dependency, but disabling incremental compilation makes no difference. Should this be filed in a different issue?

```
RandomInsano@chip-04:~/Documents/Code/pscontroller-rs/src$ rustc --version
rustc 1.26.0-nightly (55c984ee5 2018-03-16)
RandomInsano@chip-04:~/Documents/Code/pscontroller-rs/src$ cargo clean
RandomInsano@chip-04:~/Documents/Code/pscontroller-rs/src$ CARGO_INCREMENTAL=0 RUST_BACKTRACE=full cargo test --verbose
   Compiling unicode-xid v0.1.0
     Running `rustc --crate-name unicode_xid /home/RandomInsano/.cargo/registry/src/github.com-1ecc6299db9ec823/unicode-xid-0.1.0/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 --cfg 'feature="default"' -C metadata=1991908c4612c1c0 -C extra-filename=-1991908c4612c1c0 --out-dir /home/RandomInsano/Documents/Code/pscontroller-rs/target/debug/deps -L dependency=/home/RandomInsano/Documents/Code/pscontroller-rs/target/debug/deps --cap-lints allow`
   Compiling cc v1.0.8
     Running `rustc --crate-name cc /home/RandomInsano/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.8/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=c045fdd855277bee -C extra-filename=-c045fdd855277bee --out-dir /home/RandomInsano/Documents/Code/pscontroller-rs/target/debug/deps -L dependency=/home/RandomInsano/Documents/Code/pscontroller-rs/target/debug/deps --cap-lints allow`
thread 'rustc' panicked at 'index out of bounds: the len is 4847291 but the index is 4847291', /checkout/src/libserialize/opaque.rs:259:21
stack backtrace:
   0: 0xb6c80667 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h93b2fefcbf6912f6
                       at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: 0xb6c4f1eb - std::sys_common::backtrace::print::h8275cc4a0228025e
                       at libstd/sys_common/backtrace.rs:71
                       at libstd/sys_common/backtrace.rs:59
   2: 0xb6c6a457 - std::panicking::default_hook::{{closure}}::ha990f106491019b1
                       at libstd/panicking.rs:207
   3: 0xb6c6a0bf - std::panicking::default_hook::h1ab2d07a5322573d
                       at libstd/panicking.rs:223
   4: 0xb5a18137 - core::ops::function::Fn::call::hb910eaef0aecb065
   5: 0xb6c6a9cb - std::panicking::rust_panic_with_hook::h06fc1d87327d7e52
                       at libstd/panicking.rs:403
   6: 0xb6c6a7f3 - std::panicking::begin_panic_fmt::h1b61dff2ddab2600
                       at libstd/panicking.rs:349
   7: 0xb6c6a74b - rust_begin_unwind
                       at libstd/panicking.rs:325
   8: 0xb6cdab6f - core::panicking::panic_fmt::h67a7a8328d952f02
                       at libcore/panicking.rs:72
   9: 0xb6cdab23 - core::panicking::panic_bounds_check::h434bc36e56d90864
                       at libcore/panicking.rs:58
  10: 0xb6494b43 - serialize::serialize::Decoder::read_seq::hfd5771196e2a4d15
  11: 0xb6488767 - serialize::serialize::Decoder::read_struct::hf10cf5868951fa01
  12: 0xb64a8c63 - <rustc_metadata::decoder::DecodeContext<'a, 'tcx> as serialize::serialize::SpecializedDecoder<rustc::mir::interpret::AllocId>>::specialized_decode::h511731b3428098a3
  13: 0xb6477efb - serialize::serialize::Decoder::read_struct::h4ac607f9629c6909
  14: 0xb64a2463 - serialize::serialize::Decoder::read_enum::hbf203143f32403f7
  15: 0xb64a1757 - serialize::serialize::Decoder::read_enum::ha59c8e856fabf996
  16: 0xb649632f - serialize::serialize::Decoder::read_enum::h0ced95c27abbb786
  17: 0xb6497f9f - serialize::serialize::Decoder::read_enum::h349e7636a72a1bf6
  18: 0xb647508f - serialize::serialize::Decoder::read_struct::h2cc6da3b554b474e
  19: 0xb642eecf - <alloc::boxed::Box<T> as serialize::serialize::Decodable>::decode::h82bbbf14f32bf6fb
  20: 0xb64a4cbf - serialize::serialize::Decoder::read_enum::hdea19f36b5c31d62
  21: 0xb648eab3 - serialize::serialize::Decoder::read_seq::h42ae64e41bddab1e
  22: 0xb64a2f1f - serialize::serialize::Decoder::read_enum::hd21ad6388c600ca6
  23: 0xb6494ea7 - serialize::serialize::Decoder::read_enum::h04162e21b57d971a
  24: 0xb6483a7f - serialize::serialize::Decoder::read_struct::hc1701c750f0aab71
  25: 0xb648ee5b - serialize::serialize::Decoder::read_seq::h56e0898483a14ca7
  26: 0xb647bb1b - serialize::serialize::Decoder::read_struct::h75818a2537dc5cf5
  27: 0xb6494107 - serialize::serialize::Decoder::read_seq::heac0b0a9d8bae61b
  28: 0xb646d153 - <rustc::mir::Mir<'tcx> as serialize::serialize::Decodable>::decode::{{closure}}::h6a7f2382a57ec903
  29: 0xb6491cd7 - serialize::serialize::Decoder::read_seq::ha5867f7bfb039a0b
  30: 0xb646d253 - <rustc::mir::Mir<'tcx> as serialize::serialize::Decodable>::decode::{{closure}}::h6a7f2382a57ec903
  31: 0xb64f734b - rustc_metadata::decoder::<impl rustc_metadata::schema::Lazy<T>>::decode::h4478f672edc8a274
  32: 0xb63ea67f - rustc_metadata::cstore_impl::provide_extern::optimized_mir::hef17fb39368def9d
  33: 0xb593f76f - rustc::ty::maps::<impl rustc::ty::maps::queries::optimized_mir<'tcx>>::compute_result::hc8153a167fd2cc80
  34: 0xb563bf23 - rustc::dep_graph::graph::DepGraph::with_task_impl::h1f17613b988fb4f7
  35: 0xb56b817b - rustc_errors::Handler::track_diagnostics::h6f907ce7a064e0ae
  36: 0xb58b8557 - rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check::h89642d6b285a296b
  37: 0xb593f7f3 - rustc::ty::maps::<impl rustc::ty::maps::queries::optimized_mir<'tcx>>::force::hb839e3233a21d5c0
  38: 0xb593fd0f - rustc::ty::maps::<impl rustc::ty::maps::queries::optimized_mir<'tcx>>::try_get::h02048f96cd2d3418
  39: 0xb5900de7 - rustc::ty::maps::TyCtxtAt::optimized_mir::hefe786359add3d19
  40: 0xb58f1ae7 - rustc::ty::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::instance_mir::h1f2f8c01d42aa611
  41: 0xb5f573cf - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  42: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  43: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  44: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  45: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  46: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  47: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  48: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  49: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  50: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  51: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  52: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  53: 0xb5f5703b - rustc_mir::monomorphize::collector::collect_items_rec::ha798000c0ba6a115
  54: 0xb5f561fb - rustc_mir::monomorphize::collector::collect_crate_mono_items::h42ba49e83cce04e8
  55: 0xb160b1f3 - rustc::util::common::time::h36e2915d760802d3
  56: 0xb157d5f3 - rustc_trans::base::collect_and_partition_translation_items::hb00fda54eb793e4e
  57: 0xb5659243 - rustc::dep_graph::graph::DepGraph::with_task_impl::h8bf405f93ff5c5e3
  58: 0xb56b4b1b - rustc_errors::Handler::track_diagnostics::h60114bf68b72fde4
  59: 0xb58c5637 - rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check::hb32546aed983fc24
  60: 0xb59b0ccf - rustc::ty::maps::<impl rustc::ty::maps::queries::collect_and_partition_translation_items<'tcx>>::force::ha60bc632a0148e93
  61: 0xb59b134f - rustc::ty::maps::<impl rustc::ty::maps::queries::collect_and_partition_translation_items<'tcx>>::try_get::had2faba087de8a66
  62: 0xb5906137 - rustc::ty::maps::TyCtxtAt::collect_and_partition_translation_items::h68ae702fb87be5ef
  63: 0xb58ff5eb - rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::collect_and_partition_translation_items::h20d5713291984f0b
  64: 0xb157be8b - rustc_trans::base::trans_crate::h1a33bde112087c12
  65: 0xb1597a5f - <rustc_trans::LlvmTransCrate as rustc_trans_utils::trans_crate::TransCrate>::trans_crate::h635b3e99aaa549d9
  66: 0xb6de4f07 - rustc::util::common::time::h181240aad27d5e8a
  67: 0xb6dc9637 - rustc_driver::driver::phase_4_translate_to_llvm::hc89bfbd9ca36d741
  68: 0xb6e4a187 - rustc_driver::driver::compile_input::{{closure}}::hfa6b2fbd25427f03
  69: 0xb6e48513 - <std::thread::local::LocalKey<T>>::with::hfe1ad4c1ecfc61bf
  70: 0xb6e4477b - <std::thread::local::LocalKey<T>>::with::h9ec517b779397ffc
  71: 0xb6eac32f - rustc::ty::context::TyCtxt::create_and_enter::ha1426d1d12735577
  72: 0xb6dc417b - rustc_driver::driver::compile_input::h738177a00b975227
  73: 0xb6e6ad4f - rustc_driver::run_compiler_impl::ha84cb5f01fd0dd97
  74: 0xb6da15d7 - syntax::with_globals::h35a730735981f17e
  75: 0xb6dee2cb - std::sys_common::backtrace::__rust_begin_short_backtrace::h941f7d5fb16dafc2
  76: 0xb6c8e81f - __rust_maybe_catch_panic
                       at libpanic_unwind/lib.rs:102
  77: 0xb6df881f - <F as alloc::boxed::FnBox<A>>::call_box::hf3c580d870bbed4a
  78: 0xb6c83abf - std::sys_common::thread::start_thread::hc9c9c8bccde31424
                       at /checkout/src/liballoc/boxed.rs:793
                       at libstd/sys_common/thread.rs:24
  79: 0xb6c8a597 - std::sys::unix::thread::Thread::new::thread_start::h86817e665e0cb8a0
                       at libstd/sys/unix/thread.rs:90

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.26.0-nightly (55c984ee5 2018-03-16) running on armv7-unknown-linux-gnueabihf

note: compiler flags: -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `cc`.

Caused by:
  process didn't exit successfully: `rustc --crate-name cc /home/RandomInsano/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.8/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=c045fdd855277bee -C extra-filename=-c045fdd855277bee --out-dir /home/RandomInsano/Documents/Code/pscontroller-rs/target/debug/deps -L dependency=/home/RandomInsano/Documents/Code/pscontroller-rs/target/debug/deps --cap-lints allow` (exit code: 101)
```

**Comment by avl** [comment]
@ExpHP I've now added the minimized savefile-derive crate into my github repro-project linked above. That is, to https://github.com/avl/rust_ice_48923 .

The crash.sh consistently provokes the crash for my computer, even with this minimized savefile-derive crate, even if I do like:

git clean -f -x -d
./crash.sh

Should be reproducible for you too, I think! Unless there's something else that differs between our systems and the issue is basically non-deterministic (affected by something like performance, file-system timestamps or whatnot :-) )

**Comment by avl** [comment]
@RandomInsano I think your issue is something different. For one thing, the stack traces are not very similar, apart from the fact that both end up with an index out of range.

**Comment by michaelwoerister** [comment]
Thank you for all your work, @avl and @ExpHP! The small reproduction should make debugging this a lot easier. It's in my queue

**Comment by michaelwoerister** [comment]
It seems that something around `CanonicalTyGoal` is broken. In the example above, the compiler calls `dropck_outlives::try_get()` two times with the following key:

```
Canonical { 
    variables: Slice([]),
    value: ParamEnvAnd { 
        param_env: ParamEnv { 
            caller_bounds: Slice([Binder(TraitPredicate(<__S as _IMPL_SERIALIZE_FOR_X::_serde::Serializer>)),
                                  Binder(TraitPredicate(<__S as std::marker::Sized>))]),
            reveal: UserFacing,
            universe: UniverseIndex(0) 
        },
        value: std::result::Result<<__S as _IMPL_SERIALIZE_FOR_X::_serde::Serializer>::Ok,
        <__S as _IMPL_SERIALIZE_FOR_X::_serde::Serializer>::Error> 
    } 
}
```
Both keys have the same stable hash and the same `Debug` representation, but different `std::hash` values. I suspect that we are hashing pointers for `std::hash` and something is not interned right? 

cc @eddyb @nikomatsakis

**Comment by michaelwoerister** [comment]
OK, I can confirm that something is wrong with interning: The `caller_bounds` slices in the two cases have different addresses although they seem to have the same content.

**Comment by avl** [comment]
Interesting! I know serde uses a clever trick to be able to use extern types in the generated trait implementation. That trick creates a symbol named  _IMPL_SERIALIZE_FOR_<struct-name> .

**Comment by michaelwoerister** [comment]
The `Substs` within the various `TraitRef`s already have a different address:

```rust
// Version 1
Canonical {
    variables: [] (addr=1, len=0),
    value: ParamEnvAnd {
        param_env: ParamEnv {
            caller_bounds: [Binder(TraitPredicate(TraitRef { def_id: DefId(12/0:230 ~ serde[43ba]::ser[0]::Serializer[0]),
                                                             substs: [__S] (addr=7f0b0e434820, len=1) })), 
                            Binder(TraitPredicate(TraitRef { def_id: DefId(2/0:886 ~ core[3da8]::marker[0]::Sized[0]),
                                                             substs: [__S] (addr=7f0b0e434820, len=1) }))] (addr=7f0b03e35410, len=2),
            reveal: UserFacing,
            universe: UniverseIndex(0) 
        },
        value: std::result::Result<TraitRef { def_id: DefId(12/0:230 ~ serde[43ba]::ser[0]::Serializer[0]), 
                                              substs: [__S] (addr=7f0b0e434820, len=1) }::Ok,
                                   TraitRef { def_id: DefId(12/0:230 ~ serde[43ba]::ser[0]::Serializer[0]),
                                              substs: [__S] (addr=7f0b0e434820, len=1) }::Error>
    }
}

// Version 2
Canonical {
    variables: [] (addr=1, len=0),
    value: ParamEnvAnd {
        param_env: ParamEnv {
            caller_bounds: [Binder(TraitPredicate(TraitRef { def_id: DefId(12/0:230 ~ serde[43ba]::ser[0]::Serializer[0]),
                                                             substs: [__S] (addr=7f0b0e434588, len=1) })), 
                            Binder(TraitPredicate(TraitRef { def_id: DefId(2/0:886 ~ core[3da8]::marker[0]::Sized[0]),
                                                             substs: [__S] (addr=7f0b0e434588, len=1) }))] (addr=7f0b03e33c28, len=2),
            reveal: UserFacing,
            universe: UniverseIndex(0) 
        },
        value: std::result::Result<TraitRef { def_id: DefId(12/0:230 ~ serde[43ba]::ser[0]::Serializer[0]), 
                                              substs: [__S] (addr=7f0b0e434588, len=1) }::Ok,
                                   TraitRef { def_id: DefId(12/0:230 ~ serde[43ba]::ser[0]::Serializer[0]),
                                              substs: [__S] (addr=7f0b0e434588, len=1) }::Error>
    }
}
```

**Comment by nikomatsakis** [comment]
@michaelwoerister hmm. That sounds bad, yes. That interning-slices code can be subtle, I guess I messed it up somehow perhaps.

**Comment by nikomatsakis** [comment]
Although, that .. doesn't *necessarily* look related to canonicalization. But it seems likely to be the fault of that code.

**Comment by michaelwoerister** [comment]
It turns out that `__S` is a `TyParam` (with index `#0`) in both cases. I'm now looking into whether the `Symbol` used for storing the name has a different interning key in both cases.

**Comment by nikomatsakis** [comment]
@michaelwoerister ok let me know if I can help

**Comment by michaelwoerister** [comment]
I think that's the problem: serde probably uses `Symbol::gensym()` (for hygiene?) and we end up with two different `Symbol` keys. `HashStable`, on the other hand, only hashes the string contents of the symbol. It does not make a difference between the two versions. The question is: how do we deal with this? What does it even mean that the names are gensymed? Do we need to find a stable way for hashing gensymed symbols?

**Comment by jasondavies** [comment]
Just ran into this issue when compiling stdsimd (incrementally) on Linux with nightly.  `CARGO_INCREMENTAL=0 cargo build` fixed the issue for me.

**Comment by avl** [comment]
I just re-ran my test case using rustc 1.27.0-nightly (ad610bed8 2018-04-11) , and the problem seems to be resolved!

I can't reproduce the problem with this newer rustc version.

Yay for progress! :-)
